packages/webpack-inline-constant-exports-plugin/index.js

```diff
@@ -0,0 +1,117 @@
+const HarmonyImportSideEffectDependency = require( 'webpack/lib/dependencies/HarmonyImportSideEffectDependency' );
+const HarmonyImportSpecifierDependency = require( 'webpack/lib/dependencies/HarmonyImportSpecifierDependency' );
+const ConstDependency = require( 'webpack/lib/dependencies/ConstDependency' );
+
+const moduleTypes = [ 'javascript/auto', 'javascript/dynamic', 'javascript/esm' ];
+
+function addConstantExport( module, name, value ) {
+	if ( ! module.constantExports ) {
+		module.constantExports = {};
+	}
+	module.constantExports[ name ] = value;
+}
+
+function removeDependency( module, dep ) {
+	module.removeDependency( dep );
+	dep.module.reasons = dep.module.reasons.filter( r => r.dependency !== dep );
+}
+
+class InlinePlugin {
+	constructor( matchers ) {
+		if ( Array.isArray( matchers ) ) {
+			this.matchers = matchers;
+		} else {
+			this.matchers = [ matchers ];
+		}
+	}
+
+	isConstantsModule( module ) {
+		return module && this.matchers.some( matcher => matcher.test( module.resource ) );
+	}
+
+	apply( compiler ) {
+		compiler.hooks.compilation.tap( 'InlinePlugin', ( compilation, { normalModuleFactory } ) => {
+			const handleParser = parser => {
+				const handleExport = ( statement, declaration ) => {
+					const { module } = parser.state;
+
+					if ( ! this.isConstantsModule( module ) ) {
+						return;
+					}
+
+					if (
+						statement.type === 'ExportNamedDeclaration' &&
+						declaration.type === 'VariableDeclaration' &&
+						declaration.kind === 'const'
+					) {
+						for ( const declarator of declaration.declarations ) {
+							if ( declarator.init.type === 'Literal' ) {
```

Comment:
Does `Literal` include things like `{}` or `[]`? If so, we'll need to add some extra logic to make sure we're not accepting any mutable types, as those break the assumptions here.

Reply:
Object and Array literals have different AST types: `ObjectExpression` and `ArrayExpression`. All uses of `Literal` seem to be primitives where different instances can still be `===` to each other: strings, numbers, booleans, `null`... If you find some instance that's not safe (astexplorer.net can help), please let me know. Answers also @sirreal's question.